src/test/java/io/confluent/idesidecar/restapi/kafkarest/RecordsV3ApiImplIT.java

```diff
@@ -0,0 +1,401 @@
+          "%s-key".formatted(topic),
+          "JSON",
+          loadResource("schemas/product.schema.json")
+      );
+      createSchema(
+          "%s-value".formatted(topic),
+          "PROTOBUF",
+          loadResource("schemas/product.proto")
+      );
+      // Schema version 1 would be created by the above calls,
+      // but the following call should fail to find version 40
+      // (I mean, who even has 40 versions of a schema?)
+      produceRecordThen(
+          null, topic, "key", keySchemaVersion, "value", valueSchemaVersion)
+          .statusCode(404)
+          .body("message", matchesRegex("^Version \\d+ not found.; error code: 40402$"));
+    }
+
+    @Test
+    void shouldThrowBadRequestIfKeyAndValueDataAreNull() {
+      var topic = randomTopicName();
+      createTopic(topic);
+      produceRecordThen(null, topic, null, null)
+          .statusCode(400)
+          .body("message", equalTo("Key and value data cannot both be null"));
+    }
+
+    private static ArgumentSets badData() {
+      return ArgumentSets
+          .argumentsForFirstParameter(
+              SchemaManager.SchemaFormat.AVRO,
+              SchemaManager.SchemaFormat.JSON
+          )
+          .argumentsForNextParameter(
+              Stream.of(
+                  Map.of(),
+                  Map.of(
+                      "id", 123
+                  ),
+                  Map.of(
+                      "id", 123,
+                      "name", 50
+                  ),
+                  Map.of(
+                      "id", "invalid",
+                      "name", "hello",
+                      "price", 123.45
+                  ),
+                  Map.of(
+                      "id", 10,
+                      "name", List.of("hello", "world"),
+                      "price", 123.45
+                  )
+              )
+          );
+    }
+
+    @CartesianTest
+    @CartesianTest.MethodFactory("badData")
+    void shouldThrowBadRequestIfSchemaIsNotCompatibleWithData(
+        SchemaManager.SchemaFormat keyFormat, Object badData
+    ) {
+      var topic = randomTopicName();
+      createTopic(topic);
+      var keySchema = createSchema(
+          "%s-key".formatted(topic),
+          keyFormat.schemaProvider().schemaType(),
+          getProductSchema(keyFormat)
+      );
+
+      produceRecordThen(
+          null, topic, badData, keySchema.getVersion(), null, null)
+          .statusCode(400)
+          .body("message", containsString("Failed to parse data"));
+
+      var valueSchema = createSchema(
+          "%s-value".formatted(topic),
+          keyFormat.schemaProvider().schemaType(),
+          getProductSchema(keyFormat)
+      );
+
+      produceRecordThen(
+          null, topic, null, null, badData, valueSchema.getVersion())
+          .statusCode(400)
+          .body("message", containsString("Failed to parse data"));
+    }
+
+    private static Stream<Arguments> shouldThrowBadRequestForInvalidProtobufData() {
```

Comment:
The name of this method is weird, because it's not throwing anything. Maybe this??
```suggestion
    /**
     * Inputs for test cases for when the (Protobuf) schema is not compatible with the data.
     * @return the arguments for the test cases
     */
    static Stream<Arguments> invalidProtobufData() {
```